torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,367 @@ static struct PyModuleDef _module = {
     -1,
     _methods};
 
+/**
+ * Stores relevant guard debug information, e.g., failure str for a LeafGuard
+ * failure. The data structure is also accessible in Python.
+ */
+struct GuardDebugInfo {
+  GuardDebugInfo(int num_guards_executed, std::string failure_reason) {
+    this->num_guards_executed = num_guards_executed;
+    this->failure_reason = failure_reason;
+  }
+
+  std::string repr() {
+    return "GuardDebugInfo(num_guards_executed=" +
+        std::to_string(num_guards_executed) +
+        ", failure_reason=" + failure_reason + ")";
+  }
+
+  // Failure reason for a leaf guard.
+  std::string failure_reason;
+
+  // Total number of executed guards so far.
+  int num_guards_executed;
+};
+
+/**
+ * Base class for the terminal or leaf guard in the GuardManager hierarchy.
+ */
+class LeafGuard {
+ public:
+  // Runs the guard and prepares a GuardDebugInfo object.
+  std::pair<bool, GuardDebugInfo> check_with_debug_info(py::object value) {
+    bool result = run_guards(value);
+    if (result == false) {
+      std::string reason = get_failure_reason(value);
+      return std::make_pair(result, GuardDebugInfo(0, reason));
+    }
+    return std::make_pair(result, GuardDebugInfo(0, "PASS"));
+  }
+
+  // Runs the guard
+  bool check(py::object value) {
+    return run_guards(value);
+  }
+
+  virtual bool run_guards(py::object value) = 0;
+  virtual std::string get_failure_reason(py::object value) = 0;
+  virtual ~LeafGuard() = default;
+};
+
+/**
+ * Represents a leaf guard that accepts the python guard check function. We
+ * would like to have most of the guards in C++ (to avoid a Python function
+ * call).  But, it will take some time to reach that goal. Also, there might be
+ * cases where its too tedious to write an equivalent C++ guard.
+ *
+ * PythonLambdaGuard allows us to gradually move to C++. We can start from all
+ * guards of type PythonLambaGuard and incrementally move expensive guards to
+ * C++.
+ */
+class PythonLambdaGuard : public LeafGuard {
```

Comment:
Runtime polymorphism in C++ is unusual, imo, compared to compile time.

If all a PythonLambdaGuard is doing is defining 2 functions, consider a template. Its kind of a personal bias, but I *strongly* prefer templated code. 

```
template <typename CheckFunc, typename FailureFunc>
class LeafGuard {
public:
    LeafGuard(CheckFunc check_fn, FailureFunc failure_fn)
        : _check_fn(check_fn), _failure_fn(failure_fn) {}

    std::pair<bool, GuardDebugInfo> check_with_debug_info(py::object value) {
        bool result = run_guards(value);
        if (!result) {
            std::string reason = get_failure_reason(value);
            return std::make_pair(result, GuardDebugInfo(0, reason));
        }
        return std::make_pair(result, GuardDebugInfo(0, "PASS"));
    }

    bool check(py::object value) {
        return run_guards(value);
    }

    bool run_guards(py::object value) {
        return _check_fn(value);
    }

    std::string get_failure_reason(py::object value) {
        return _failure_fn(value);
    }

private:
    CheckFunc _check_fn;
    FailureFunc _failure_fn;
};

auto check_lambda = [](py::object value) -> bool { /* ... */ };
auto failure_lambda = [](py::object value) -> std::string { /* ... */ };

LeafGuard<decltype(check_lambda), decltype(failure_lambda)> guard(check_lambda, failure_lambda);
```

Reply:
SGTM, author's choice :)